Allow specifying zip archive member order.

dhasenan opened the pull request:
Fixes https://issues.dlang.org/show_bug.cgi?id=16310

wilzbach commented on std/zip.d:
```diff
@@ -376,7 +385,9 @@ final class ZipArchive
     void[] build()
-    {   uint i;
+    {
+        import std.algorithm : sort;
```
it's better to be more selective as this will avoid the need to parse all `std.algorithm` packages.
-> `std.algorithm.sorting : sort`;

wilzbach commented on std/zip.d:
```diff
@@ -385,7 +396,8 @@ final class ZipArchive
         uint archiveSize = 0;
         uint directorySize = 0;
-        foreach (ArchiveMember de; _directory)
+        auto directory = _directory.values().sort!((x, y) => x.index < y.index);
```
use `.release` at the end of the chain to get the array back

wilzbach commented on std/zip.d:
```diff
@@ -874,6 +887,40 @@ debug(print)
 {
+    import std.random : uniform;
+    import std.conv : to;
+    import std.algorithm : canFind;
```
`std.algorithm.searching`

wilzbach commented on std/zip.d:
```diff
@@ -874,6 +887,40 @@ debug(print)
+    string[] names;
+    foreach (i; 0..20)
+    {
+        auto name = uniform(0, 0xFFFF).to!string;
```
please don't use unseeded randomness in tests.

```d
import std.random : Mt19937;
auto gen = Mt19937(42);
```

dhasenan commented:
Randomized testing with a different seed each time would provide better coverage, but in this case, it would only ensure that (a) we're not randomly getting a pre-sorted array and (b) std.algorithm.sorting.sort works correctly. So I used a fixed seed as you recommended and also added output so you can see the exact sequence that we're expecting.

I also adjusted the array generation so it doesn't need canFind.

DmitryOlshansky commented:
LGTM

wilzbach commented on std/zip.d:
```diff
@@ -385,7 +396,8 @@ final class ZipArchive
         uint archiveSize = 0;
         uint directorySize = 0;
-        foreach (ArchiveMember de; _directory)
+        auto directory = _directory.values().sort!((x, y) => x.index < y.index).release;
```
Will this yield to an noticeable overhead if no specific order is needed?

andralex commented:
Auto-merge toggled on
